src/spec/datatypes.rs

```diff
@@ -0,0 +1,388 @@
+        }
+    }
+}
+
+fn deserialize_decimal<'de, D>(deserializer: D) -> Result<PrimitiveType, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    let (precision, scale) = s
+        .trim_start_matches(r"decimal(")
+        .trim_end_matches(')')
+        .split_once(',')
+        .ok_or_else(|| D::Error::custom("Decimal requires precision and scale: {s}"))?;
+
+    Ok(PrimitiveType::Decimal {
+        precision: precision.parse().map_err(D::Error::custom)?,
+        scale: scale.parse().map_err(D::Error::custom)?,
+    })
+}
+
+fn serialize_decimal<S>(precision: &u32, scale: &u32, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    serializer.serialize_str(&format!("decimal({precision},{scale})"))
+}
+
+fn deserialize_fixed<'de, D>(deserializer: D) -> Result<PrimitiveType, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let fixed = String::deserialize(deserializer)?
+        .trim_start_matches(r"fixed[")
+        .trim_end_matches(']')
+        .to_owned();
+
+    fixed
+        .parse()
+        .map(PrimitiveType::Fixed)
+        .map_err(D::Error::custom)
+}
+
+fn serialize_fixed<S>(value: &u64, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    serializer.serialize_str(&format!("fixed[{value}]"))
+}
+
+impl fmt::Display for PrimitiveType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            PrimitiveType::Boolean => write!(f, "boolean"),
+            PrimitiveType::Int => write!(f, "int"),
+            PrimitiveType::Long => write!(f, "long"),
+            PrimitiveType::Float => write!(f, "float"),
+            PrimitiveType::Double => write!(f, "double"),
+            PrimitiveType::Decimal {
+                precision: _,
+                scale: _,
+            } => write!(f, "decimal"),
+            PrimitiveType::Date => write!(f, "date"),
+            PrimitiveType::Time => write!(f, "time"),
+            PrimitiveType::Timestamp => write!(f, "timestamp"),
+            PrimitiveType::Timestampz => write!(f, "timestampz"),
+            PrimitiveType::String => write!(f, "string"),
+            PrimitiveType::Uuid => write!(f, "uuid"),
+            PrimitiveType::Fixed(_) => write!(f, "fixed"),
+            PrimitiveType::Binary => write!(f, "binary"),
+        }
+    }
+}
+
+/// DataType for a specific struct
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+#[serde(rename = "struct", tag = "type")]
+pub struct StructType {
+    /// Struct fields
+    pub fields: Vec<StructField>,
+}
+
+impl StructType {
+    /// Get structfield with certain id
+    pub fn get(&self, index: usize) -> Option<&StructField> {
+        self.fields.iter().find(|field| field.id as usize == index)
+    }
+    /// Get structfield with certain name
+    pub fn get_name(&self, name: &str) -> Option<&StructField> {
+        self.fields.iter().find(|field| field.name == name)
+    }
+}
```

Comment:
In Python and Java we also have convenient formatting of the struct. I think in rust this is [the `Display` trait](https://doc.rust-lang.org/std/fmt/trait.Display.html). A struct gets formatted as `struct<string, int, long, double>` based on the fields. Maybe also something to consider for rust

Reply:
I'ce included it in this PR. How is the formatting for maps and lists?

Reply:
Map: `map<string, double>`
List: `list<string>`

Reply:
Can you add this in a separate PR? This way we can get this one in.